### PR TITLE
Search Price Filtering Feature

### PR DESCRIPTION
**Pull request by Hreherch**
You can now filter by price when searching by location or keyword. 

Solves: #140

### CODE REVIEW

File: app/src/androidTest/java/comcmput301f16t01/github/carrier/ElasticUserControllerTest.java [hunk 1]
```diff
@@ -1,14 +1,14 @@
 package comcmput301f16t01.github.carrier;
 
-import java.util.Random;
-
 /**
  * Created by kiete on 11/13/2016.
  */
```
Comment (sehoven) [inline app/src/androidTest/java/comcmput301f16t01/github/carrier/ElasticUserControllerTest.java:5]: This auto-generated Javadoc should be removed

File: app/src/main/java/comcmput301f16t01/github/carrier/MainActivity.java [hunk 2]
```diff
@@ -355,42 +365,92 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
+                        }
+                    });
+
+                    // Perform the async update with the listener in place to stop the refresh
+                    // symbol when the async tasks have finished.
+                    RequestController rc = new RequestController();
+                    rc.performAsyncUpdate();
+                }
+            });
+
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/MainActivity.java:432]: This is so cool I love this.

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/Request.java [hunk 3]
```diff
@@ -150,8 +157,10 @@ public String getId() {
 
     @Override
     public String toString() {
-        String requestAsString = "Request From: " + rider.getUsername() + "\n";
-        requestAsString += "Description: " + description;
+        String requestAsString = "Request From: " + rider.getUsername() + "\n" +
+                "Description: " + description + "\n" +
+                "Price: " + getFare() + "\n" +
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/Request.java]: Do FareCalculator fc = new FareCalculator(); fc.toString(getFare()); to get a properly formatted fare (right now it is the fare * 100)

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/Request.java [hunk 4]
```diff
@@ -150,8 +157,10 @@ public String getId() {
...
+                "Price per KM: " + (fare/100)/distance;
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/Request.java]: Round to some number of decimal points? This gives a HUGE float right now.

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 5]
```diff
@@ -221,18 +271,53 @@ public RequestList fetchAllRequestsWhereRider( User rider ) {
...
+        RequestList filteredRequests = new RequestList();
+        for ( Request request : searchResult ) {
+            double pricePerKM = request.getFare() / request.getDistance();
+            // ensure the price per kilometer is greater than the specified minimum
+            if ( pricePerKM < minPricePerKM * 100 ) { continue; }
+            // ensure the price per kilometer is less than the specified maximum.
+            if ( maxPricePerKM != null && maxPricePerKM * 100 < pricePerKM ) { continue; }
+            filteredRequests.add( request ); // add the request if it is in range
+        }
+        searchResult.replaceList( filteredRequests );
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:336]: If I filter for a price per KM that is minimum 2, it returns a request with a price per KM of 1.951981... I assume that this is possibly a rounding error?

Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:336]: So I also tried this same case with a maximum of 2 (minimum of 0) and it didn't return that request (with a price per KM of 1.951981...)

Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:336]: There is a request with a price per KM of 1.04744539... This one works as expected if I use a maximum or minimum of 1. If I use a minimum of 1.05 (maximum of 2), this request is still returned. If I use a maximum of 1.05 (minimum of 0), the request is not returned. Again, possibly a rounding error somewhere? How precise do we want to be?

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java [hunk 6]
```diff
@@ -221,18 +271,53 @@ public RequestList fetchAllRequestsWhereRider( User rider ) {
...
             e.printStackTrace();
         }
-        requestList.replaceList( foundRequests );
+        requestsWhereRider.replaceList( foundRequests );
         return foundRequests;
     }
 
-    public void addListener( Listener listener ) {
-        requestList.addListener( listener );
+    public void performAsyncUpdate() {
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:278]: Javadocs?

File: app/src/main/java/comcmput301f16t01/github/carrier/Searching/SearchActivity.java [hunk 7]
```diff
@@ -59,14 +88,134 @@ public void onClick(DialogInterface dialogInterface, int i) {
...
+        startActivity(intent);
+    }
+
+    /**
+     * We create a bundle to put in the intent for SearchResultActivity so the activity can make proper
+     * calls and get the requests filtered by the user's need.
+     * @param intent the intent for the SearchResultActivity
+     * @see SearchResultsActivity
+     */
+    private void bundleFilters(Intent intent) {
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/Searching/SearchActivity.java:126]: I like this method a lot.

File: app/src/main/res/values/colors.xml [hunk 8]
```diff
@@ -2,7 +2,7 @@
 <resources>
     <color name="colorPrimary">#424242</color>
     <color name="colorPrimaryDark">#212121</color>
-    <color name="colorAccent">#FF4081</color>
+    <color name="colorAccent">#7e57c2</color>
```
Comment (sehoven) [inline app/src/main/res/values/colors.xml]: AHA! YOU ARE THE CULPRIT THAT GOT RID OF THE HOT PINK.

**Review by sehoven — CHANGES_REQUESTED** [review]
I think because you pulled a branch or two there's A LOT going on here so I focussed my attention on the filtering part. I did some skimming over the parts that didn't seem relevant so if someone else had time to also skim that would probably be beneficial.

File: app/src/main/res/values/colors.xml (hunk 8 above)
Comment (MandyMeindersma) [reply app/src/main/res/values/colors.xml]: WE NEED HOT PINK!!!!
@kieter @Hreherch @MichaelParadis

File: app/src/main/java/comcmput301f16t01/github/carrier/Searching/SearchResultsActivity.java [hunk 9]
```diff
@@ -27,19 +27,56 @@ protected void onCreate(Bundle savedInstanceState) {
```
Comment (sehoven) [inline app/src/main/java/comcmput301f16t01/github/carrier/Searching/SearchResultsActivity.java:28]: I realize that this is kind of outside of the scope of your branch but could you do me a favour and add this **setTitle("Search Results");** under setContentView(...). There is a typo in the title in the main branch.

Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/Searching/SearchResultsActivity.java:28]: I'm sorry if this is hard to see from here. In the SearchResultsActivity.java file in the OnCreate method under the setContentView line.

File: app/src/androidTest/java/comcmput301f16t01/github/carrier/ElasticUserControllerTest.java (hunk 1 above)
Comment (Hreherch) [reply app/src/androidTest/java/comcmput301f16t01/github/carrier/ElasticUserControllerTest.java:5]: @kieter

File: app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java (hunk 6 above)
Comment (Hreherch) [reply app/src/main/java/comcmput301f16t01/github/carrier/Requests/RequestController.java:278]: solved in another branch

**Comment by Hreherch** [comment]
@sehoven Would love a quick look over :)

**Comment by sehoven** [comment]
I put in 2 requests (user is shoven, keyword for both requests is periwinkle if you want to try to duplicate). I'm trying to filter by price. One has a price of $11.60, one has a price of $18.41. If I do a min of 0 and a max of 11.61, only the first is returned. Great. If I do a min of 11.61 with no max, no requests are returned (oh no!). If I do a min of 11.61 with a max of 20, the second request is returned (yay!). So something is up with the optional max perhaps.

**Comment by sehoven** [comment]
For price by kilometer (using same examples as above), the first is 1.25, the second is 1.75. With any valid min and no max, no requests are returned. With a min of 1.22 and a max of 5, both requests are returned, bump min up to 1.23 and only the second is returned. With a min of 1.26 and a max of 1.67, I can still see the second request (the second request again has a value of 1.75), bump it down to 1.66 and it does what it's supposed to - I can't see the second request anymore.

**Comment by Hreherch** [comment]
I left in float values when I converted to double, should be maybe okay now? @sehoven

**Review by sehoven — APPROVED** [review]

**Pull request merged**